### PR TITLE
Fix #369, replace findParentDir.sync with Promise

### LINKED ISSUE #369: Performance: TSLint - Replace `findParentDir.sync` in `getLintConfig` with async call / promise
While looking at some performance issues, I saw that `findParentDir` is called synchronously in tslint - need to refactor that to be asynchronous and return a promise instead.

The file is `vim\core\oni-plugin-tslint\lib\index.js`. This is called whenever we decide to do linting, which right now is on buffer enter and buffer save.

### PR DESCRIPTION
**Pull request by keforbes**
`find-parent-dir` has a dependency on a deprecated call (https://github.com/thlorenz/find-parent-dir/issues/4) so I decided to replace it with [find-up](https://github.com/sindresorhus/find-up), which does basically the same thing but has support for Promises.  Also, `find-up` returns the full path of the filename rather than the parent dir.  This reduced some code since every usage of `find-parent-dir` was `join`ing the directory name and filename.

Now, with that said, I don't know anything about Promises.  I tried to emulate the Promise code found elsewhere in Oni but it's very likely I made a rookie mistake.  Please take a look and let me know if you have any suggestions.

### CODE REVIEW

**Comment by extr0py** [comment]
Cool, it does look like that library is a bit more convenient in terms of syntax for Oni.

> This reduced some code since every usage of find-parent-dir was joining the directory name and filename.

Excellent!

> Now, with that said, I don't know anything about Promises. I tried to emulate the Promise code found elsewhere in Oni but it's very likely I made a rookie mistake. 

Yes, Promises are a little tough to understand at first...  These are a couple decent references:
https://developers.google.com/web/fundamentals/getting-started/primers/promises
https://davidwalsh.name/promises
http://www.mattgreer.org/articles/promises-in-wicked-detail/

The cool thing is that they can be chained, which avoids having tons of nested callbacks. It seems like handling asychronous actions in JavaScript started with callbacks, then moved towards Promises, and is now moving towards async/await: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function

File: browser/src/ProjectConfig.ts [hunk 1]
```diff
@@ -39,13 +39,13 @@ const DefaultConfiguration: IProjectConfiguration = {
  * Search upward for the relevant .oni folder
  */
 export function getProjectConfiguration(filePath: string): Q.Promise<IProjectConfiguration> {
-    const oniDir = findParentDir.sync(filePath, ".oni")
+    const oniDir = findUp.sync(".oni", { cwd: filePath })
```
Comment (extr0py) [inline browser/src/ProjectConfig.ts]: Seems like we could actually make this `async` now (which is the non-blocking code-path). Something like:
```
return findUp(".oni", {  cwd: filePath })
.then((oniDir) => {
      if (!oniDir) {
          return DefaultConfiguration
      }
      return loadConfigurationFromFolder(oniDir)
})
```

**Comment by extr0py** [comment]
Added a couple comments about the Promises - we can actually simplify and remove the `Q.defer` and just return the values directly. Let me know if you have any questions!

**Comment by keforbes** [comment]
Ok, I made that change to `ProjectConfig.ts`, let me know if there are any other changes needed.

**Comment by extr0py** [comment]
Thanks for making all the changes, @keforbes ! I just had one other request to see if we can use the async version of `findUp` in tslint - since the promises are all set there, it should be an easy switch, and helpful for performance.

**Comment by keforbes** [comment]
> see if we can use the async version of findUp in tslint

I might not be understanding this.  Are you suggesting I do:
```
 .then(() => {
    findUp("tsconfig.json", { cwd: currentWorkingDirectory })
 })
.then((project) => {
```

Given this call of `findUp` returns another promise, I think I'd have to nest the remaining promises like this:
```
 .then(() => {
   findUp("tsconfig.json", { cwd: currentWorkingDirectory })
    .then((project) => {
      //all other code
    }
 })
```

Let me know if I'm missing something.

**Comment by extr0py** [comment]
Close - you can actually chain promises. Your code would work, it's just missing the `return` statement, so if it looks like this:

```
 .then(() => {
    return findUp("tsconfig.json", { cwd: currentWorkingDirectory })
 })
.then((project) => {
```

This is nice because you can eliminate that nesting (which happens a lot with callbacks). 

Then, there is an even shorter version, because if you use a lambda and omit the `{` and `}`, the `return` is implied. So this form actually means the same as the one above... just a bit more concise:
``` 
.then(() => findUp("tsconfig.json", { cwd: currentWorkingDirectory }))
.then((project) => {
```

Thanks!

**Comment by keforbes** [comment]
Cool, thanks for explaining it @extr0py!

**Comment by extr0py** [comment]
Sure thing! Thanks for taking this on, @keforbes , and addressing all th e feedback. I'll bring this in - it helps move us to leverage Promises (and hopefully, in the longer run, `async` and `await`).

**Pull request merged**